node/src/types/sync_leap.rs

```diff
@@ -290,20 +295,1676 @@ impl FetchItem for SyncLeap {
+            &switch_blocks,
+            query,
+            &trusted_ancestor_headers,
+            &signed_block_headers,
+            add_proofs,
+        );
+
+        // `sync_leap` is a well formed SyncLeap structure for the test chain. We can use the blocks
+        // it contains to generate SyncLeap structures as required for the test, because we know the
+        // heights of the blocks in the test chain as well as their sigs.
+        let highest_block = sync_leap.signed_block_headers.last().unwrap().clone();
+        let lowest_blocks: Vec<_> = sync_leap
+            .trusted_ancestor_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .collect();
+        let middle_blocks: Vec<_> = sync_leap
+            .signed_block_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .map(|block_header_with_metadata| block_header_with_metadata.block_header)
+            .collect();
+        let sync_leap = SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header: lowest_blocks.first().unwrap().clone(),
+            trusted_ancestor_headers: middle_blocks,
+            signed_block_headers: vec![highest_block],
+        };
+        assert!(sync_leap.highest_block_header_and_signatures().1.is_some());
+    }
+
+    #[test]
+    fn should_not_return_sigs_when_highest_block_is_not_signed() {
+        // Chain
+        // 0   1   2   3   4   5   6   7   8   9   10   11
+        // S           S           S           S
+        let switch_blocks = [0, 3, 6, 9];
+
+        let mut rng = TestRng::new();
+
+        let query = 5;
+        let trusted_ancestor_headers = [4, 3];
+        let signed_block_headers = [6, 9, 11];
+        let add_proofs = true;
+        let sync_leap = make_test_sync_leap(
+            &mut rng,
+            &switch_blocks,
+            query,
+            &trusted_ancestor_headers,
+            &signed_block_headers,
+            add_proofs,
+        );
+
+        // `sync_leap` is a well formed SyncLeap structure for the test chain. We can use the blocks
+        // it contains to generate SyncLeap structures as required for the test, because we know the
+        // heights of the blocks in the test chain as well as their sigs.
+        let highest_block = sync_leap.signed_block_headers.last().unwrap().clone();
+        let lowest_blocks: Vec<_> = sync_leap
+            .trusted_ancestor_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .collect();
+        let middle_blocks: Vec<_> = sync_leap
+            .signed_block_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .collect();
+        let sync_leap = SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header: highest_block.block_header,
+            trusted_ancestor_headers: lowest_blocks,
+            signed_block_headers: middle_blocks,
+        };
```

Comment:
Similar to above, if we construct the initial `sync_leap` with no signed block headers, it will be valid and would allow us to remove all this code.

Reply:
I'd rather keep this as is, because if we create a `SyncLeap` without `signed_block_headers` we're not going to cover [this line](https://github.com/casper-network/casper-node/blob/dev/node/src/types/sync_leap.rs#L167) with the test.

Since the iteration on switch blocks will not happen, I can, for example, change `==` to `!=` in the implementation and the test will still pass.

Reply:
As discussed, we'll keep as is for now.